src/wallet/wallet.cpp

```diff
@@ -121,9 +121,15 @@ bool CWallet::AddKeyPubKey(const CKey& secret, const CPubKey &pubkey)
     if (!fFileBacked)
         return true;
     if (!IsCrypted()) {
-        return CWalletDB(strWalletFile).WriteKey(pubkey,
-                                                 secret.GetPrivKey(),
-                                                 mapKeyMetadata[pubkey.GetID()]);
+        if (pwalletdb) {
```

Comment:
I don't like the code duplication here at every call site, as well as the manual deleting setting-to-NULL of pwalletdb. What about a inner class for context management like
```
/** Database reference. Can be nested, in which case the outer reference will own the handle.
  */
class DBRef
{
private:
    bool owning;
public:
    DBRef(CWallet * wallet): owning(false)
    {
        if (!wallet->pwalletdb) {
            wallet->pwalletdb = new CWalletDB(wallet->strWalletFile);
            owning = true;
       }
    }
    CWalletDB &get() { return *wallet->pwalletdb;  }
    ~DBRef()
    {
         if(owning) {
             delete wallet->pwalletdb;
             wallet->pwalletdb = NULL;
         }
    }
};
```
Then you can change this to
```
DBRef ref;
ref.get().WriteKey(pubkey, secret.GetPrivKey(), mapKeyMetadata[pubkey.GetID());
```